### PR TITLE
fix(file source): correct an error with line aggregation in `continue_through` and `halt_before`

### LINKED ISSUE #3237: Multiline
I have trouble using multiline in the file source. The first occurance of the multiline event is detected correctly, but following occurences are merged into the preceding line, even though they do not match the start_pattern.

vector.toml:
```data_dir = "./vector-data"

[sources.logfiles]
include = ["test-files/test3.log"]
type = "file"
start_at_beginning = true
# Multiline

[sources.logfiles.multiline]
condition_pattern = "^\\s"
mode = "continue_through"
start_pattern = "^\\s"
timeout_ms = 10

[sinks.bar]
inputs = ["logfiles"]
type = "console"
encoding.codec = "json"
```

logfile.log:
```
{"level":30,"time":"2020-07-28T09:08:56.984Z", "msg": "test"}
 first multiline
 test
{"level":30,"time":"2020-07-28T09:08:56.984Z", "msg": "test"}
 second multiline
 test
{"level":30,"time":"2020-07-28T09:08:56.984Z", "msg": "test"}
 test
 test
{"level":30,"time":"2020-07-28T09:08:56.984Z", "msg": "test"}
```

Output:
```
Jul 28 17:34:51.725  INFO vector: Log level "info" is enabled.
Jul 28 17:34:51.727  INFO vector: Loading configs. path=["vector.toml"]
Jul 28 17:34:51.731  INFO vector: Vector is starting. version="0.10.0" git_version="" released="Mon, 27 Jul 2020 09:51:50 +0000" arch="x86_64"
Jul 28 17:34:51.731  INFO vector::topology: Running healthchecks.
Jul 28 17:34:51.731  INFO vector::topology: Starting source "logfiles"
Jul 28 17:34:51.732  INFO vector::topology: Starting sink "bar"
Jul 28 17:34:51.732  INFO vector::topology::builder: Healthcheck: Passed.
Jul 28 17:34:51.732  INFO source{name=logfiles type=file}: vector::sources::file: Starting file server. include=["logfile.log"] exclude=[]
Jul 28 17:34:51.734  INFO source{name=logfiles type=file}:file_server: file_source::file_server: Found file to watch. path="logfile.log" file_position=0
{"file":"logfile.log","host":"thinkpad-jan","message":"{\"level\":30,\"time\":\"2020-07-28T09:08:56.984Z\", \"msg\": \"test\"}","source_type":"file","timestamp":"2020-07-28T15:34:51.739358461Z"}
{"file":"logfile.log","host":"thinkpad-jan","message":" first multiline\n test","source_type":"file","timestamp":"2020-07-28T15:34:51.739447421Z"}
{"file":"logfile.log","host":"thinkpad-jan","message":"{\"level\":30,\"time\":\"2020-07-28T09:08:56.984Z\", \"msg\": \"test\"}\n second multiline\n test","source_type":"file","timestamp":"2020-07-28T15:34:51.739453560Z"}
{"file":"logfile.log","host":"thinkpad-jan","message":"{\"level\":30,\"time\":\"2020-07-28T09:08:56.984Z\", \"msg\": \"test\"}\n test\n test","source_type":"file","timestamp":"2020-07-28T15:34:51.739457476Z"}
{"file":"logfile.log","host":"thinkpad-jan","message":"{\"level\":30,\"time\":\"2020-07-28T09:08:56.984Z\", \"msg\": \"test\"}","source_type":"file","timestamp":"2020-07-28T15:34:51.751723805Z"}
```

### PR DESCRIPTION
**Pull request by MOZGIII**
Closes #3237.

### CODE REVIEW

File: src/sources/file/line_agg.rs [hunk 1]
```diff
@@ -140,6 +151,20 @@ where
...
+                        let val = match val {
+                            // If we have to emit just one line - that's easy,
+                            // we just return it.
+                            (Emit::One(val), src) => (val, src),
+                            // If we have to emit two lines - take the second
+                            // one and stash it, then return the first one.
+                            // This way, the stashed line will be returned
+                            // on the next stream poll.
+                            (Emit::Two(val, to_stash), src) => {
+                                debug_assert!(self.stashed.is_none());
```
Comment (bruceg) [inline src/sources/file/line_agg.rs:163]: Might want to comment why this should be true.

File: src/sources/file/line_agg.rs [hunk 2]
```diff
@@ -425,6 +459,59 @@ mod tests {
...
+            mode: Mode::HaltBefore,
+            timeout: Duration::from_millis(10),
+        };
+        let expected = vec![
+            "merged 1\nmerged 2",
+            " splitter 1",
+            "merged 3\nmerged 4",
+            " splitter 2",
+        ];
+        run_and_assert(&lines, config, &expected);
```
[the remark this answers is not in the document]
Comment (MOZGIII) [reply src/sources/file/line_agg.rs:512]: Is seems to be reasonable to test the start and condition patterns elsewhere. The implementation details are such that start pattern and condition pattern are covered well by the rest of the tests, so it’s not clear to what degree should we test those in this case.

Were you thinking about just switching from `` to something like `^merged`?

[the remark this answers is not in the document]
Comment (MOZGIII) [reply src/sources/file/line_agg.rs:512]: I see. You're right, the coverage should be improved. I'm not sure if 16 is the comprehensive amount of test cases to cover all the states though. How about addressing it in another PR? This one looks like it does it's the job and solves the issue at hand.

[the remark this answers is not in the document]
Comment (bruceg) [reply src/sources/file/line_agg.rs:515]: Ok

[the remark this answers is not in the document]
Comment (MOZGIII) [reply src/sources/file/line_agg.rs:515]: I created https://github.com/timberio/vector/issues/3272 so this doesn't get lost.

**Comment by MOZGIII** [comment]
I gave the tests another look before the merge, and found a set of serious issues. :sweat_smile:
They're now fixed, and tests are adapted accordingly.  @bruceg, could you please take another look?
The relevant commit is https://github.com/timberio/vector/pull/3262/commits/22386ec93bc0f11afd7b2661bd9b022746992b9e.

File: src/sources/file/line_agg.rs [hunk 3]
```diff
@@ -293,6 +293,7 @@ fn add_next_line(buffered: &mut BytesMut, line: Bytes) {
 #[cfg(test)]
 mod tests {
     use super::*;
+    use pretty_assertions::assert_eq;
```
Comment (bruceg) [inline src/sources/file/line_agg.rs:311]: Interesting, I'll have to remember that one for my own tests. Too badd we can't turn that on across all the sources at once.

Comment (MOZGIII) [reply src/sources/file/line_agg.rs:311]: I noticed it being used at the `file/mod.rs` and it looked really good.

Can't we import it at `/lib.rs` with an old `extern crate pretty_assertions` + macro use syntax?

**Pull request merged**